### ISSUE TITLE
Problem with performance on voc dataset

### ISSUE THREAD

**Issue by heurainbow** [andy-yun/pytorch-0.4-yolov3#20]
I trained the yolov2 on voc dataset and got following performance. It seems not correct for such a high prec and low recall.  I also tried marvis repo https://github.com/marvis/pytorch-yolo2 and the result is reasonable.
```
python train.py -d cfg/voc.data -c cfg/yolo-voc.cfg -w darknet19_448.conv.23 -l
```
```
2018-09-13 16:29:17 [020] correct: 1045, precision: 0.914261, recall: 0.086866, fscore: 0.158656
2018-09-13 16:55:40 [030] correct: 2319, precision: 0.938866, recall: 0.192768, fscore: 0.319859
2018-09-13 17:22:24 [040] correct: 2666, precision: 0.906494, recall: 0.221613, fscore: 0.356152
2018-09-13 17:48:53 [050] correct: 3214, precision: 0.913068, recall: 0.267165, fscore: 0.413373
2018-09-13 18:15:42 [060] correct: 3986, precision: 0.893922, recall: 0.331338, fscore: 0.483470
2018-09-13 18:42:38 [070] correct: 3731, precision: 0.916257, recall: 0.310141, fscore: 0.463417
2018-09-13 19:10:27 [080] correct: 4692, precision: 0.918379, recall: 0.390025, fscore: 0.547519
2018-09-13 19:38:13 [090] correct: 3848, precision: 0.910123, recall: 0.319867, fscore: 0.473363
2018-09-13 20:05:11 [100] correct: 5322, precision: 0.904640, recall: 0.442394, fscore: 0.594201
2018-09-13 20:32:11 [110] correct: 4921, precision: 0.891809, recall: 0.409061, fscore: 0.560857
2018-09-13 20:59:11 [120] correct: 4925, precision: 0.904666, recall: 0.409393, fscore: 0.563690
2018-09-13 21:26:28 [130] correct: 5215, precision: 0.876765, recall: 0.433500, fscore: 0.580149
2018-09-13 21:53:38 [140] correct: 4876, precision: 0.899631, recall: 0.405320, fscore: 0.558850
2018-09-13 22:21:21 [150] correct: 5108, precision: 0.904551, recall: 0.424605, fscore: 0.577922
2018-09-13 22:48:52 [160] correct: 5569, precision: 0.913250, recall: 0.462926, fscore: 0.614404
2018-09-13 23:15:48 [170] correct: 5630, precision: 0.914406, recall: 0.467997, fscore: 0.619119
2018-09-13 23:42:43 [180] correct: 5571, precision: 0.915982, recall: 0.463092, fscore: 0.615168
2018-09-14 00:09:45 [190] correct: 5675, precision: 0.911940, recall: 0.471737, fscore: 0.621811
2018-09-14 00:37:38 [200] correct: 5744, precision: 0.912325, recall: 0.477473, fscore: 0.626864
2018-09-14 01:04:43 [210] correct: 5623, precision: 0.921803, recall: 0.467415, fscore: 0.620293
2018-09-14 01:31:52 [220] correct: 5607, precision: 0.917976, recall: 0.466085, fscore: 0.618256
2018-09-14 01:59:03 [230] correct: 5739, precision: 0.917066, recall: 0.477057, fscore: 0.627620
2018-09-14 02:26:00 [240] correct: 5806, precision: 0.913468, recall: 0.482627, fscore: 0.631563
2018-09-14 02:53:08 [250] correct: 5703, precision: 0.919839, recall: 0.474065, fscore: 0.625667
2018-09-14 03:20:08 [260] correct: 5728, precision: 0.918243, recall: 0.476143, fscore: 0.627103
2018-09-14 03:47:23 [270] correct: 5675, precision: 0.921865, recall: 0.471737, fscore: 0.624102
2018-09-14 04:14:34 [280] correct: 5668, precision: 0.924331, recall: 0.471155, fscore: 0.624156
2018-09-14 04:41:38 [290] correct: 5762, precision: 0.919713, recall: 0.478969, fscore: 0.629894
2018-09-14 05:08:53 [300] correct: 5717, precision: 0.918689, recall: 0.475229, fscore: 0.626413
2018-09-14 05:35:59 [310] correct: 5802, precision: 0.914277, recall: 0.482294, fscore: 0.631471
2018-09-14 06:45:14 [020] correct: 0, precision: 0.000000, recall: 0.000000, fscore: 0.000000

```

**Comment by lianuo** [comment]
@heurainbow Hi,could you share some image result for your test? both andy's and marvis', thank you~

**Comment by heurainbow** [comment]
Different mAP, and I also found the test process of this repo is much slower than that of marvis.
### marvis

```
AP for aeroplane = 0.7845
AP for bicycle = 0.7892
AP for bird = 0.7218
AP for boat = 0.6306
AP for bottle = 0.4629
AP for bus = 0.7770
AP for car = 0.7967
AP for cat = 0.8668
AP for chair = 0.5217
AP for cow = 0.7667
AP for diningtable = 0.7069
AP for dog = 0.8227
AP for horse = 0.8016
AP for motorbike = 0.7797
AP for person = 0.7494
AP for pottedplant = 0.4726
AP for sheep = 0.7477
AP for sofa = 0.7253
AP for train = 0.8643
AP for tvmonitor = 0.7088
Mean AP = 0.7248
```
![predictions](https://user-images.githubusercontent.com/6753276/45528187-ce4e6880-b811-11e8-8730-041a44bb5493.jpg)

### andy
```
AP for aeroplane = 0.7569
AP for bicycle = 0.7686
AP for bird = 0.7062
AP for boat = 0.5735
AP for bottle = 0.4282
AP for bus = 0.7653
AP for car = 0.7887
AP for cat = 0.8342
AP for chair = 0.4882
AP for cow = 0.7154
AP for diningtable = 0.6685
AP for dog = 0.8124
AP for horse = 0.7960
AP for motorbike = 0.7480
AP for person = 0.7226
AP for pottedplant = 0.4645
AP for sheep = 0.6970
AP for sofa = 0.6677
AP for train = 0.8283
AP for tvmonitor = 0.6957
Mean AP = 0.6963
```
andy's produce no boxes on dog.jpg.

I wonder something is wrong with the detect.py and andy's model.

**Comment by andy-yun** [comment]
Thanks for the detailed information.
My version and marvis version are different at loss function weights and some codes.
So compare the performance between the two systems after over 1,000 epochs.
Maybe the results are different for the same training epochs.
In the experimental results, the local maximum results do not guarantee the best performance.

By the way, on my dataset, I tuned the loss weight on region_layer.py. I did not test the code on voc-data.
Therefore, if you use some different data, you may adjust the loss weights and object_scale in yolo-voc.cfg.

You can inspect the code in region_layer.py.

        coord_mask, conf_mask = coord_mask.view(cls_anchor_dim).to(self.device), conf_mask.sqrt().to(self.device)

        t3 = time.time()
        loss_coord = self.coord_scale * nn.MSELoss(size_average=False)(coord*coord_mask, tcoord*coord_mask)/2
        # sqrt(object_scale)/2 is almost equal to 1.
        loss_conf = nn.MSELoss(size_average=False)(conf*conf_mask, tconf*conf_mask)/2 
        loss_cls = self.class_scale * nn.CrossEntropyLoss(size_average=False)(cls, tcls) if cls.size(0) > 0 else 0

**Comment by lianuo** [comment]
@andy-yun Hi,Andy, have you complete the training on your two GPU ,could your share the prediction result of images?

**Comment by ydixon** [comment]
Familiar faces here :D. @andy-yun Thanks for the repo. Out of all the recent repo's of Pytorch yolo-v3, this one seems to converge the best. Might I ask a few question:

1. Why use mseLoss on confidence instead of bceloss (logistic regression)?
2. Why do mseloss on sigmoid(tx) instead of tx?
3. Why do crossEntropyLoss instead of bceloss on loss_cls? (following the paper, but I could imagine CELoss will work as well considering the results here)

@lianuo So far I found repos that do bceloss on coordinates loss converge pretty bad for large datasets, I modified my loss function to use mseloss, it's converging way better

**Comment by andy-yun** [comment]
@lianuo  not yet. As you reported, the performance seems lower than the original C version when I check the mAP on 44 epochs.
But, I would like to postpone my decision at least 200 epoch. In the previous experiment with my own data, I did train my model over 500,000 epoch about 1 week on C version. Current epoch on pytorch version is just 56. T_T. I don't know how long to train a large dataset model, such as COCO.

@heurainbow I think you should train the model until the intermittent results shows some balanced result between precision and recall even though the epoch may be greater than @marvis version.

@ydixon The loss function has been borrowed intact in the @marvis version. But, I modified the weight to each loss. I think they are good choices as @lianuo  pointed.

**Comment by andy-yun** [comment]
@heurainbow I modified the data augmentation routine and fixed the acnhor box repetition routine. 
At the anchorbox repetition (expand the dimention), I misunderstood the dimension information. 
After correct that, I realized that the performance is similar to marvis and the original C version. Would you check the performance?

**Comment by andy-yun** [comment]
@heurainbow  The followings are the result:

history:
2018-10-12 16:09:51 [010] correct: 8038, precision: 0.360885, recall: 0.668163, fscore: 0.468643
2018-10-12 17:25:41 [020] correct: 8399, precision: 0.421827, recall: 0.698171, fscore: 0.525902
2018-10-12 18:39:01 [030] correct: 8536, precision: 0.401354, recall: 0.709559, fscore: 0.512699
2018-10-12 19:52:40 [040] correct: 8615, precision: 0.443729, recall: 0.716126, fscore: 0.547936
2018-10-12 21:06:43 [050] correct: 8824, precision: 0.500652, recall: 0.733500, fscore: 0.595106
2018-10-12 22:21:55 [060] correct: 8837, precision: 0.516935, recall: 0.734580, fscore: 0.606828
2018-10-12 23:37:15 [070] correct: 8825, precision: 0.522004, recall: 0.733583, fscore: 0.609962
2018-10-13 00:53:36 [080] correct: 8809, precision: 0.512241, recall: 0.732253, fscore: 0.602794
2018-10-13 02:10:16 [090] correct: 8863, precision: 0.517004, recall: 0.736741, fscore: 0.607612
2018-10-13 03:27:03 [100] correct: 8849, precision: 0.522991, recall: 0.735578, fscore: 0.611325
2018-10-13 04:43:16 [110] correct: 8844, precision: 0.523965, recall: 0.735162, fscore: 0.611846
2018-10-13 05:59:03 [120] correct: 8817, precision: 0.529391, recall: 0.732918, fscore: 0.614742
2018-10-13 07:14:37 [130] correct: 8826, precision: 0.526705, recall: 0.733666, fscore: 0.613189

mAP:
AP for aeroplane  = 0.7323 with  285 views
AP for bicycle    = 0.8226 with  337 views
AP for bird       = 0.7346 with  459 views
AP for boat       = 0.6042 with  263 views
AP for bottle     = 0.4164 with  469 views
AP for bus        = 0.8058 with  213 views
AP for car        = 0.8108 with 1201 views
AP for cat        = 0.8765 with  358 views
AP for chair      = 0.5087 with  756 views
AP for cow        = 0.7436 with  244 views
AP for diningtable = 0.7468 with  206 views
AP for dog        = 0.8347 with  489 views
AP for horse      = 0.8398 with  348 views
AP for motorbike  = 0.7959 with  325 views
AP for person     = 0.7387 with 4528 views
AP for pottedplant = 0.4664 with  480 views
AP for sheep      = 0.7739 with  242 views
AP for sofa       = 0.7765 with  239 views
AP for train      = 0.8566 with  282 views
AP for tvmonitor  = 0.7263 with  308 views
Mean AP = 0.7306 with total 12032 views
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           Results:
------------------------------
aeroplane       0.732
bicycle         0.823
bird            0.735
boat            0.604
bottle          0.416
bus             0.806
car             0.811
cat             0.877
chair           0.509
cow             0.744
diningtable     0.747
dog             0.835
horse           0.840
motorbike       0.796
person          0.739
pottedplant     0.466
sheep           0.774
sofa            0.777
train           0.857
tvmonitor       0.726
==============================
 Average        0.731
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

detected results:
![predictions](https://user-images.githubusercontent.com/39820731/46896757-e1f4f980-ceb9-11e8-8dba-ae48c6ccff0c.jpg)

**Comment by andy-yun** [comment]
@heurainbow you can change the weights.. Then bounding boxes are corrected.
I think you would change the combinations of object_scale = 1.25, coord_scale = 0.5, class_scale=1. 
@marvis hard coded the scales without changing of yolov-voc.cfg.

**Comment by andy-yun** [comment]
@heurainbow when i set the threshold as follows, the performances are different according to thresholds.
coord_scale=1, object_scale=1, class_scale=1 mAP=50.4
coord_scale=1, object_scale=3, class_scale=2 mAP=72.8
So, if you use different thresholds, the performances are varied.
In these experiments, I did not consider focal_loss and bceloss.
